### PR TITLE
Canonical module type and type

### PR DESCRIPTION
**Pull request by jonludlam**
Allows @canonical tags to be put on module type and types.

### CODE REVIEW

File: src/model/error.mli [hunk 1]
```diff
@@ -1,5 +1,9 @@
 type t
 
+type 'a with_warnings = { value : 'a; warnings : t list }
+
+type warning_accumulator
+
```
Comment (Julow) [inline src/model/error.mli:6]: Why these changes ?

**Review by Julow — APPROVED** [review]
I think we should name the new constructors `` `CanonicalType `` and `` `CanonicalModuleType `` instead. (also `T` and `Ty` are not consistent with other names we have, they should be `MT` and `T`)

File: src/model/error.mli (hunk 1 above)
Comment (jonludlam) [reply src/model/error.mli:6]: Oops, I think I meandered off into 'separate out the parser library' territory while doing this work, and didn't quite revert everything!

**Comment by jonludlam** [comment]
Mmm, I think you're right about the names of the new constructors. Thanks for the feedback!
